src/NHibernate/Type/GenericSetType.cs

```diff
@@ -76,5 +77,20 @@ public override object Instantiate(int anticipatedSize)
 		{
 			return new HashSet<T>();
 		}
+
+		protected override bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
```

Comment:
[`ISet.SetEquals`](https://msdn.microsoft.com/en-us/library/dd412096(v=vs.110).aspx) does not require its argument to be a set. It would be slightly more efficient to rebuild target as a list instead of as a set.

(There is a `SetEquals` on `CollectionHelper` from `NHibernate.Util` namespace too. A bit redundant with .Net `ISet.SetEquals`, but NHibernate is quite older than sets in .Net.)